### ISSUE TITLE
Steam Freezing while installing games.

### ISSUE THREAD

**Issue by Redbatman89** [ValveSoftware/steam-for-linux#7720]
So I think ever since the update for the February 13th release on Tuesday March 23rd may have screwed up my client a bit. Now when I install a game or update the steam client will freeze for a minute randomly then go back to being functional. The games still work and I can play them. But its getting annoying that this is happening now.

I have tried shutting down and powering up my computer, installing the new Nvidia Drivers. But that didn't work so I'm thinking its a client issue. Is this happening to anyone else? I'm on POP OS 20.04

**Comment by class101** [comment]
I note this behavior too on Arch Linux, not sure if it is hardware related or Steam client side issue, it started recently but before February probably  because I'm using Steam client beta

What I note is that it seems to happen when Steam uses disk access after downloading. For me the symptoms encountered are even worse because it produces intermittent freezes of some part of the desktop, the mouse continue to work but nothing is clickable for a short period, very weird behavior I never experienced before.

It's as if Steam was asking for a higher priority than the kernel and everything Steam does on disk takes priority over vital system functions, such as displaying desktop etc etc. 

```
OS: Arch Linux x86_64 
Kernel: 5.11.9-zen1-1-zen 
Uptime: 8 hours, 30 mins 
Packages: 1280 (pacman) 
Shell: bash 5.1.4 
Resolution: 1920x1080 
DE: Cinnamon 4.8.6 
WM: Mutter (Muffin) 
WM Theme: Mojave-dark (Mojave-dark) 
Theme: Mojave-dark [GTK2/3] 
Icons: McMojave-circle [GTK2/3] 
Terminal: guake 
CPU: Intel i7-4770K (8) @ 3.900GHz 
GPU: NVIDIA GeForce GTX 980 
GPU: NVIDIA GeForce GTX 980 
Memory: 4143MiB / 15937MiB
Disk: Samsung SSD 850 Pro
```

**Comment by kisak-valve** [comment]
Most likely newer symptoms of #4978.

**Comment by Redbatman89** [comment]
So I did some testing and at first I thought it was only native Linux games giving me the issue since I had downloaded a Windows game too and see what would happen. At First no freezing for one game, so then I tried another and this however it froze a bit. (I play on proton sometimes) So then next I had downloaded a game I pulled up system monitor and to watch for any high cpu usage. Sure enough there was a spike on one of the cores for 100%, and its an Intel I7 4970K quad core. 

Thought this may help. Steam has had other issues like this in the past when a few months ago they had to disable shader cache for Nvidia cards due to the client not liking Nvidia Drivers past 440.100 for awhile.

**Comment by class101** [comment]
I took a screenshot this morning of my task manager with a steamwebhelper process automatically set to a CPU nice of -11, higher than everything on the system, higher than pulseaudio defaulting to cpu nice -10.

![steamwebhelper](https://user-images.githubusercontent.com/1810484/112714944-4d41b300-8edd-11eb-9261-8cc829c501d3.png)

> Most likely newer symptoms of #4978.

Nothing to do with this 4 year old issue.

**Comment by Redbatman89** [comment]
A little acknowledgement would be nice Steam.

**Comment by gbschenkel** [comment]
Hi, using Slackware here. I am having this problem a long time. Got a 15MB update from Satisfactory and it got me 3 minutes just to install it while freezing/hanging the machine. I was running Steam on Console, and got some throughput:

```
No cached sticky mapping in ActivateActionSet.Installing breakpad exception handler for appid(steam)/version(1618256785)
Installing breakpad exception handler for appid(steam)/version(1618256785)
Installing breakpad exception handler for appid(steam)/version(1618256785)
ExecuteSteamURL: "steam://open/minigameslist"
saving roaming config store to 'sharedconfig.vdf'
roaming config store 2 saved successfully
Installing breakpad exception handler for appid(steam)/version(1618256785)
Installing breakpad exception handler for appid(steam)/version(1618256785)
Installing breakpad exception handler for appid(steam)/version(1618256785)
Fossilize INFO: Autogroup scheduling is not enabled on this kernel. Will rely entirely on nice().
Installing breakpad exception handler for appid(steam)/version(1618256785)
Installing breakpad exception handler for appid(steam)/version(1618256785)
Installing breakpad exception handler for appid(steam)/version(1618256785)
Installing breakpad exception handler for appid(steam)/version(1618256785)
Installing breakpad exception handler for appid(steam)/version(1618256785)
Installing breakpad exception handler for appid(steam)/version(1618256785)
Installing breakpad exception handler for appid(steam)/version(1618256785)
Installing breakpad exception handler for appid(steam)/version(1618256785)
Installing breakpad exception handler for appid(steam)/version(1618256785)
Installing breakpad exception handler for appid(steam)/version(1618256785)
Installing breakpad exception handler for appid(steam)/version(1618256785)
Installing breakpad exception handler for appid(steam)/version(1618256785)
Installing breakpad exception handler for appid(steam)/version(1618256785)
Installing breakpad exception handler for appid(steam)/version(1618256785)
Installing breakpad exception handler for appid(steam)/version(1618256785)
Installing breakpad exception handler for appid(steam)/version(1618256785)
Installing breakpad exception handler for appid(steam)/version(1618256785)
Installing breakpad exception handler for appid(steam)/version(1618256785)
Installing breakpad exception handler for appid(steam)/version(1618256785)
Installing breakpad exception handler for appid(steam)/version(1618256785)
Installing breakpad exception handler for appid(steam)/version(1618256785)
[0420/030653.529455:INFO:crash_reporting.cc(270)] Crash reporting enabled for process: renderer
ExecuteSteamURL: "steam://open/downloads"
Work thread 'CDepotReconstruct - Decrypt Thr' is marked exited, but we could not immediately join prior to deleting -- proceeding without join
Installing breakpad exception handler for appid(steam)/version(1618256785)
Installing breakpad exception handler for appid(steam)/version(1618256785)
Installing breakpad exception handler for appid(steam)/version(1618256785)
Installing breakpad exception handler for appid(steam)/version(1618256785)
Fossilize INFO: Autogroup scheduling is not enabled on this kernel. Will rely entirely on nice().
Installing breakpad exception handler for appid(steam)/version(1618256785)
Installing breakpad exception handler for appid(steam)/version(1618256785)
Installing breakpad exception handler for appid(steam)/version(1618256785)
Installing breakpad exception handler for appid(steam)/version(1618256785)
Installing breakpad exception handler for appid(steam)/version(1618256785)
Installing breakpad exception handler for appid(steam)/version(1618256785)
Installing breakpad exception handler for appid(steam)/version(1618256785)
Installing breakpad exception handler for appid(steam)/version(1618256785)
Installing breakpad exception handler for appid(steam)/version(1618256785)
Installing breakpad exception handler for appid(steam)/version(1618256785)
Installing breakpad exception handler for appid(steam)/version(1618256785)
Installing breakpad exception handler for appid(steam)/version(1618256785)
Installing breakpad exception handler for appid(steam)/version(1618256785)
Installing breakpad exception handler for appid(steam)/version(1618256785)
Installing breakpad exception handler for appid(steam)/version(1618256785)
Installing breakpad exception handler for appid(steam)/version(1618256785)
Installing breakpad exception handler for appid(steam)/version(1618256785)
Installing breakpad exception handler for appid(steam)/version(1618256785)
Installing breakpad exception handler for appid(steam)/version(1618256785)
Installing breakpad exception handler for appid(steam)/version(1618256785)
Installing breakpad exception handler for appid(steam)/version(1618256785)
Installing breakpad exception handler for appid(steam)/version(1618256785)
Installing breakpad exception handler for appid(steam)/version(1618256785)
Installing breakpad exception handler for appid(steam)/version(1618256785)
Installing breakpad exception handler for appid(steam)/version(1618256785)
ExecuteSteamURL: "steam://open/minigameslist"
saving roaming config store to 'sharedconfig.vdf'
roaming config store 2 saved successfully
Installing breakpad exception handler for appid(steam)/version(1618256785)
Installing breakpad exception handler for appid(steam)/version(1618256785)
Installing breakpad exception handler for appid(steam)/version(1618256785)
Installing breakpad exception handler for appid(steam)/version(1618256785)
Installing breakpad exception handler for appid(steam)/version(1618256785)
Installing breakpad exception handler for appid(steam)/version(1618256785)
Installing breakpad exception handler for appid(steam)/version(1618256785)
Installing breakpad exception handler for appid(steam)/version(1618256785)
Installing breakpad exception handler for appid(steam)/version(1618256785)
Installing breakpad exception handler for appid(steam)/version(1618256785)
Installing breakpad exception handler for appid(steam)/version(1618256785)
Installing breakpad exception handler for appid(steam)/version(1618256785)
ExecuteSteamURL: "steam://open/largegameslist"
ExecuteSteamURL: "steam://open/downloads"
Installing breakpad exception handler for appid(steam)/version(1618256785)
ExecuteSteamURL: "steam://open/minigameslist"
saving roaming config store to 'sharedconfig.vdf'
roaming config store 2 saved successfully
[2021-04-20 03:16:51] uninstalled manifest found in /home/schenkel/.local/share/Steam/package/steam_client_publicbeta_ubuntu12 (1).
```
I think the problem is with [Fossilize](https://github.com/ValveSoftware/Fossilize) doing something while decompressing the downloaded files.

Found this while writing this post. https://ask.csdn.net/questions/3196333

**Comment by Redbatman89** [comment]
Just wanted to say the the Update for April 12th did nothing to change the situation any better.

**Comment by class101** [comment]
I think I somehow fixed my problem here after reading some SSD recommandations for performance.

I realized that a Linux service, fstrim, which by default is configured to TRIM drives every week to maintain good read-write performance, was not running on my system at all (technically, that's normal because my drives are encrypted in LVM / LUKS, and for security reasons it is not enabled by default and you have to enable this manually to maintain the best performance over time)

Good article [here](https://easylinuxtipsproject.blogspot.com/p/ssd.html#ID6)


Before my write speed was down to a crap speed as low at 70-120 mb/s

As it had been a long time since I had reinstalled my system, I did fresh reinstall and with all ssd performance recommandations applied this time, and I'm back to my expected 500mb/s in read and write (not latest ssd generation, old samsung ssd 850 pro but still good)

So I don't think I will find this problem in a long time.

@Redbatman89 

Might be worth trying if your system can be Trimed like that

```bash
sudo fstrim /
```
If it says 'not supported' or something like, then it is not enabled

and to test the write speed of your drive where Steam games are installed

Write speed test
```bash
cd /path/to/drive/where/steam/common
dd if=/dev/zero of=benchmark.test bs=2G count=1 oflag=dsync
```

**Comment by J0hannaR** [comment]
I was just about to report this bug, then I found this issue here, which is possibly the same one.

Tl:dr: Steam is downloading faster than the page cache in RAM gets written onto disk. When the cache is full, Linux switches from non-blocking I/O to blocking I/O, which causes the system to freeze.

Solution: Steam could limit the download speed to file I/O rate and do a regular fsync or some sort of direct write/own caching procedure when downloading/updating apps.

I discovered that when updating Cyberpunk 2077. Download (1.5GB) was fine, but when it's at 100% the system freezes in like 2 minutes intervals and becomes uncontrollable. After reboot, Steam started updating again, but this time a 30 GB update, maybe due to corruption. After the update is at 100%, same behavior.

Reinstalling went fine, but with a similar but different behavior. Steam downloads for half a minute at about 120MB/s, but then the rate goes to 0, system starts freezing again and disk I/O stays high and decreases slowly to 0. Then download goes up again. 

The thing is, the update always ends up in a permanent freeze, while re-downloading the whole game worked in the end. My guess is, the 1.5 GB was downloaded completely into my RAM*, and when the update mechanisms starts to decompress and extract the files, page cache gets full, Linux kernel switches to synchronous, blocking IO and freezes. Maybe when doing a full reinstallation the decompressing/extracting already starts while downloading, therefore making it 'bearable', while updating does it after download completion, which impedes the issue.

*with standard manjaro setting, the page cache is 20% of RAM ( /proc/sys/vm/dirty_ratio ), and the interval for flushing to disk is 15 seconds ( /proc/sys/vm/dirty_writeback_centisecs ). So with my 120MB/s download rate and 32GB RAM the update may be completed in RAM only.

So as a temporary solution, reducing my download speed manually in my Steam settings to 20-40 MB/s helped, also playing around with sysctl settings vm.dirty_*  . Also activating BFQ I/O may help (https://www.reddit.com/r/Fedora/comments/ay7dkh/linux_large_transfers_freeze_system_high_io/), but I didn't test it.

**Comment by class101** [comment]
@J0hannaR 

Have you tested your drive write speed is ok ? Because the symptoms you are describing were the same I got but that's was when my drive was slower than a normal hdd.

Test this, (not in /tmp because it is usually a tmpfs into ram memory)

```
cd /path/to/drive/where/steam/common
dd if=/dev/zero of=benchmark.test bs=2G count=1 oflag=dsync
```
If you are around 100 Mb/s then yes you will freeze a lot because it is slow.

**Comment by J0hannaR** [comment]
> Have you tested your drive write speed is ok ? Because the symptoms you are describing were the same I got but that's was when my drive was slower than a normal hdd.

Yes, I did some intensive benchmarking in the past. So I already know that my old SanDisk SSD Plus is really slow. It's also my fully LUKS encrypted system drive, which certainly doesn't help.

Nevertheless I checked again with dd:

With /dev/zero I get a peak write speed of 340MB/s with a blocksize of 256MB, going higher or lower with blocksize ends up with 120MB/s. I also tested /dev/random, which I think is more similar to compressed data, giving also write rates of 120 MB/s, no matter the block size.
When setting the CPU governor to powersave, which I usually have, it even goes further down to 50 MB/s (even though cryptsetup benchmark reports about 2 GB/s for aes-xts with 512B key length).

That's all similar results to my benchmarks in Windows with CrystalDiskMark and AES containers, so yes it's definitively the SSD that's bottlenecking and Steam doesn't seem to take into account that can happen. Admittedly, my setup with a gigabit connection (which is pretty much the best you can get in Germany) and an incredibly slow SSD is a pretty weird and rare combination, I guess.

The same issue when downloading does happen from time to time on my unencrypted storage SSD (Crucial P1 512GB) too though. dd writing from /dev/zero is about 800 MB/s and /dev/random 260 MB/s, so it's technically fast enough. It also is much less severe, since it doesn't freeze my system completely and is a much shorter interval, it also doesn't happen everytime. But those gaps, in which my desktop environment slows down for about 10 seconds, can occur too:
![gitsteam](https://user-images.githubusercontent.com/83412556/116685298-65ac5e00-a9b2-11eb-8c98-03f6db991e41.png)

Maybe due to some background stuff happening, I don't know.

**Comment by Redbatman89** [comment]
So latest steam update yesterday still hasn't fixed this for me, Im thinking it has something to do with Steam still. Also for trim, POP OS automatically has trim enabled by default. i checked and it was already enabled.

**Comment by Redbatman89** [comment]
Still no fix after June 7th update.

**Comment by class101** [comment]
> Also for trim, POP OS automatically has trim enabled by default

Just a side note, every Linux distro has it enabled by default for unencrypted partition yes by the service `fstrim.timer`

But it is not enabled for encrypted volume by default.

I discovered that by doing a simple `sudo fstrim -av`,  it was only trimming `/boot` because it was the only part I left unencrypted.

```
/boot: 1,7 GiB (1801662464 bytes) trimmed on /dev/sda1
```

I needed to :

- add `issue_discards = 1` to `/etc/lvm/lvm.conf` 
- append `:allow-discards` to the cryptdevice boot entry as below
```
options nvidia-drm.modeset=1 intel_pstate=active cryptdevice=UUID=68b25e3a-89fb-4a1b-96b9-8baf3ed62a58:cryptroot:allow-discards root=/dev/vg0/root
```

and now fstrim works for all mount points as / is encrypted.

```
/boot: 1,7 GiB (1801662464 bytes) trimmed on /dev/sda1
/mnt/sdb: 0 B (0 bytes) trimmed on /dev/sdb1
/: 395,4 MiB (414576640 bytes) trimmed on /dev/mapper/vg0-root
```

Your problem might be somewhere else but with encrypted volumes it is really important to verify this otherwise your system will never trim and I'm almost sure that's why my drive was down to a crap speed ranging from 50-70Mb/s , now it is always 550 Mb/s as expected and the Steam issue is no more really an issue here.

**Comment by kgorol** [comment]
I am literally unable to use Steam right now, as installing a new game almost instantly freezes my Steam instance.

If that matters, I am trying to install Titanfall 2 on HDD with high network speed.

**Comment by userofryzen** [comment]
I'm having this problem too with manjaro stable. It happens since a few updates but until today I have not been able to research a bit. So.. pls, valve, make something. xD I can't almost play because of this. Each time I launch steam my all pc freezes making imposible to do nothing, the problem start after finish downloading too.

**Comment by gbschenkel** [comment]
Since this is still a issue for some distribution, on Slackware I was able to convince Patrick('the boss') to enable on default in kernel [see here](https://www.linuxquestions.org/questions/slackware-14/requests-for-current-14-2-15-0-a-4175620463/page497.html#post6242963) and [here](https://www.linuxquestions.org/questions/slackware-14/requests-for-current-14-2-15-0-a-4175620463/page585.html#post6295401). Then in Slackware 15.0 we got autogroup enabled on kernel by default.

```
Automatic process group scheduling (SCHED_AUTOGROUP)

CONFIG_SCHED_AUTOGROUP:

This option optimizes the scheduler for common desktop workloads by
automatically creating and populating task groups. This separation
of workloads isolates aggressive CPU burners (like build jobs) from
desktop applications. Task group autogeneration is currently based
upon task session.

Symbol: SCHED_AUTOGROUP [=y]
Type : bool
Defined at init/Kconfig:1205
Prompt: Automatic process group scheduling
Location:
-> General setup
Selects: CGROUPS [=y] && CGROUP_SCHED [=y] && FAIR_GROUP_SCHED [=y]
```
Like my post  before https://github.com/ValveSoftware/steam-for-linux/issues/7720#issuecomment-823010382, fossilize try use a kernel feature, which is not enabled doing all the trouble you guys are experiencing. First this is not a Steam issue, is your distro issue, you guys need enable the kernel feature 'autogroup' some how on your distro, recompiling the kernel, or grabbing another kernel version.

**Comment by gbschenkel** [comment]
I also like to advice to you guys(users) to enable the `PREEMPT_VOLUNTARY=y` on kernel.
```
Voluntary Kernel Preemption (Desktop)
          his option reduces the latency of the kernel by adding more
	  "explicit preemption points" to the kernel code. These new
	  preemption points have been selected to reduce the maximum
	  latency of rescheduling, providing faster application reactions,
	  at the cost of slightly lower throughput.

	  This allows reaction to interactive events by allowing a
	  low priority process to voluntarily preempt itself even if it
	  is in kernel mode executing a system call. This allows
	  applications to run more 'smoothly' even when the system is
	  under load.

	  Select this if you are building a kernel for a desktop system.
```
If your kernel is compiled with `CONFIG_PREEMPT=Y` and `PREEMPT_DYNAMIC=y`, both enabled, you can enable the PREEMPT_VOLUNTARY just putting the `preempt=voluntary` on lilo / grub / elilo default boot parameter:

Example using Grub:
Edit the `/etc/default/grub` and append as shown below the flag `preempt=voluntary`.
`GRUB_CMDLINE_LINUX_DEFAULT="preempt=voluntary"`

For Steam employees, please try add this information to the documentation of give a warning when using Steam without this and Autogroup enabled, mainly when Autogroup is not enabled. Preempt help me with issues with low input response on controllers.

**Comment by userofryzen** [comment]
@gbschenkel In my distro autogrups is enabled, manjaro. And preempt dynamic too. 
The only thing that is not enable by default I think is the voluntary thing, I'll test it..
But anyways steam was freezing my pc  with those things enabled in kernel...

**Comment by gbschenkel** [comment]
> @gbschenkel In my distro autogrups is enabled, manjaro. And preempt dynamic too. The only thing that is not enable by default I think is the voluntary thing, I'll test it.. But anyways steam was freezing my pc with those things enabled in kernel...

Try execute Steam from shell, wait for it stop, after all that `Installing breakpad exception handler for appid`.
Try do some deleting/installing the games which do some the freezing, I hope you have some little size game to test it, check on the shell what Steam is trowing out.

In my case, Fossilize before enabling Autogroups was writing on shell:
`Fossilize INFO: Autogroup scheduling is not enabled on this kernel. Will rely entirely on nice().`
after enable it change to 
`Fossilize INFO: Setting autogroup scheduling.`
and the problem stopped.

**Comment by userofryzen** [comment]
> > @gbschenkel In my distro autogrups is enabled, manjaro. And preempt dynamic too. The only thing that is not enable by default I think is the voluntary thing, I'll test it.. But anyways steam was freezing my pc with those things enabled in kernel...
> 
> Try execute Steam from shell, wait for it stop, after all that `Installing breakpad exception handler for appid`. Try do some deleting/installing the games which do some the freezing, I hope you have some little size game to test it, check on the shell what Steam is trowing out.
> 
> In my case, Fossilize before enabling Autogroups was writing on shell: `Fossilize INFO: Autogroup scheduling is not enabled on this kernel. Will rely entirely on nice().` after enable it change to `Fossilize INFO: Setting autogroup scheduling.` and the problem stopped.

nope, autogroup was enabled yesterday and another one when I had the problems so there has to be another problem.

**Comment by gbschenkel** [comment]
Well I think is need to check all kernel config, to find out why this is happening, I am adding the default Slackware kernel config, you may do a diff to find out what is different.
https://pastebin.com/AgjUe2Fg

**Comment by nathub** [comment]
I'm encountering severe slowdown across my whol system, when a Steam download finishes. Only when it finishes (like right after the notification pings up). 

I'm on openSuse but have encountered this before on Fedora and Ubuntu. But there the solution I think was to switch to Wayland (can't now). Not sure though that was a year ago.

**Comment by DisruptionSystemsINC** [comment]
Either this is still an issue, or there has been a regression. 
Since like 2 updates ago, Steam has returned to gobbling up all my M2's bandwidth and freezing up the system.

**Comment by xamindar** [comment]
I have been plagued with this issue for months.  The main culprit for me is Diablo IV updates, but many games cause it as well.
I found the following to greatly help mitigate the issue (typing this while patching Diablo IV ;) ):
Note, this depends on the amount of ram you have as these are percentage values. I suggest experimenting with different values. I have 64GB RAM for reference and I may tweak them more to test.
`sysctl -w vm.dirty_background_ratio=2`
`sysctl -w vm.dirty_ratio=5`
Defaults are 10 and 20 respectively.
Make them permanent by creating a file under etc, for example "/etc/sysctl.d/80-gamecompatibility.conf" or whatever you want to call it.

I also changed my io scheduler for my games drive to bfq, but I don't think that mattered much since we are using fast NVME here:
`echo bfq > /sys/block/nvme1n1/queue/scheduler`

**Comment by gbschenkel** [comment]
@xamindar 
I think this is making the worst, the problem is when the cache is fully, which is on RAM it need to write it on DISK, which hang until is all written, lowing the value will make it write it more often, and harm the disk, since you are making it write synchronously instead of asynchronously.

Your values are too low, I think is better to use:
```
vm.dirty_ratio=40
vm.dirty_background_ratio=5
```
for a generic use.

The best is calculate how fast your SSD/NVME works and set the value of `vm.dirty_background_bytes` instead changing `vm.dirty_background_ratio`.

References:
https://docs.oracle.com/en/database/other-databases/nosql-database/25.3/admin/linux-page-cache-tuning.html
https://linuxvox.com/blog/linux-page-cache/
https://vps.do/linux-disk-caching/
https://lonesysadmin.net/2013/12/22/better-linux-disk-caching-performance-vm-dirty_ratio/

**Comment by gbschenkel** [comment]
I have tested my system with this config:
```
vm.dirty_background_bytes = 104857600
vm.dirty_ratio = 40
vm.dirty_expire_centisecs = 1000
vm.swappiness = 10
```
And appears it is having a good performance, I tested write 10GiB with `fio`

```bash
# fio --name=read-benchmark --ioengine=libaio --iodepth=64 --rw=read --bs=4k --size=10G --numjobs=1 --runtime=60 --group_reporting
read-benchmark: (g=0): rw=read, bs=(R) 4096B-4096B, (W) 4096B-4096B, (T) 4096B-4096B, ioengine=libaio, iodepth=64
fio-3.41
Starting 1 process
read-benchmark: Laying out IO file (1 file / 10240MiB)
Jobs: 1 (f=1): [R(1)][100.0%][r=630MiB/s][r=161k IOPS][eta 00m:00s]
read-benchmark: (groupid=0, jobs=1): err= 0: pid=13821: Tue Jan 13 17:37:46 2026
  read: IOPS=253k, BW=989MiB/s (1037MB/s)(10.0GiB/10350msec)
    slat (nsec): min=982, max=73011k, avg=2796.95, stdev=50318.86
    clat (usec): min=2, max=73290, avg=238.67, stdev=393.13
     lat (usec): min=4, max=73291, avg=241.46, stdev=396.11
    clat percentiles (usec):
     |  1.00th=[  141],  5.00th=[  151], 10.00th=[  161], 20.00th=[  169],
     | 30.00th=[  180], 40.00th=[  190], 50.00th=[  194], 60.00th=[  204],
     | 70.00th=[  223], 80.00th=[  285], 90.00th=[  392], 95.00th=[  465],
     | 99.00th=[  529], 99.50th=[  586], 99.90th=[ 2147], 99.95th=[ 2868],
     | 99.99th=[ 6521]
   bw (  KiB/s): min=27112, max=1486632, per=100.00%, avg=1022727.60, stdev=358550.27, samples=20
   iops        : min= 6778, max=371658, avg=255681.60, stdev=89637.29, samples=20
  lat (usec)   : 4=0.01%, 10=0.01%, 20=0.01%, 50=0.01%, 100=0.01%
  lat (usec)   : 250=74.64%, 500=22.88%, 750=2.24%, 1000=0.04%
  lat (msec)   : 2=0.09%, 4=0.08%, 10=0.02%, 20=0.01%, 100=0.01%
  cpu          : usr=30.11%, sys=53.45%, ctx=11323, majf=0, minf=71
  IO depths    : 1=0.1%, 2=0.1%, 4=0.1%, 8=0.1%, 16=0.1%, 32=0.1%, >=64=100.0%
     submit    : 0=0.0%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, 64=0.0%, >=64=0.0%
     complete  : 0=0.0%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, 64=0.1%, >=64=0.0%
     issued rwts: total=2621440,0,0,0 short=0,0,0,0 dropped=0,0,0,0
     latency   : target=0, window=0, percentile=100.00%, depth=64

Run status group 0 (all jobs):
   READ: bw=989MiB/s (1037MB/s), 989MiB/s-989MiB/s (1037MB/s-1037MB/s), io=10.0GiB (10.7GB), run=10350-10350msec

Disk stats (read/write):
  nvme0n1: ios=41561/1317, sectors=20770240/96728, merge=0/0, ticks=8623/1627, in_queue=10384, util=50.16%
```

**Comment by gbschenkel** [comment]
I have changed, I could improve a little better with this config
```
vm.dirty_background_bytes=268435456
vm.dirty_ratio=40
vm.dirty_expire_centisecs=1000
```

## Kernel Settings Tested

| Setting | A | B | C |
|---------|---|---|---|
| vm.dirty_background_bytes | 256 MB | 100 MB | 256 MB |
| vm.dirty_ratio | 40 | 40 | 40 |
| vm.dirty_expire_centisecs | 3000 (30s) | 1000 (10s) | 1000 (10s) |

## Performance Results

| Metric | A | B | C | Winner |
|--------|---|---|---|--------|
| IOPS | 264k | 253k | 263k | A ≈ C |
| Bandwidth | 1032 MiB/s | 989 MiB/s | 1026 MiB/s | A ≈ C |
| Avg Latency | 235 µs | 241 µs | 224 µs | **C** |
| p99 | 553 µs | 529 µs | 570 µs | B |
| p99.9 | 2.2 ms | 2.1 ms | 742 µs | **C** |
| p99.99 | 9.7 ms | 6.5 ms | 2.7 ms | **C** |
| CPU (usr+sys) | 81% | 83.5% | 82% | Similar |
| Disk util | 55.5% | 50.2% | 53.9% | A |
| Write I/Os | 409 | 1317 | 81 | **C** |
| Write sectors | 11,672 | 96,728 | 2,304 | **C** |

I am using a Corsair M.2 NVMe 1TB Gen4 but on Gen3 slot.

**Comment by xamindar** [comment]
> and harm the disk, since you are making it write synchronously instead of asynchronously.

This is a non-issue for an nvme drive, there is no harm. The lower values cause it to write back to the drive synchronously about 3.2GB at a time (with my numbers) instead of a massive 12GB, which while doing it synchronously all other writes are blocked, causing the freezes. Lower amounts of data are able to be written back without causing significant freezes. This is also recommended for other high data writing processes such as databases. 

The downside is we are sacrificing speed, but that is hardly an issue with (in my case) fast DDR5 RAM and fast multicore CPU.

I am going to also experiment with a very high setting for dirty_ratio, like 80 or 90 percent, which will cause it to use almost all my available ram for caching. If the entire install/patch fits in that cache, I will be fine. But if it still breaks past that limit I can see it locking up again until the 40G+, or whatever, ends up getting written back to disk. 

Fun times tuning Linux!

**Comment by xamindar** [comment]
> The best is calculate how fast your SSD/NVME works and set the value of `vm.dirty_background_bytes` instead changing `vm.dirty_background_ratio`.

This is a good tip for calculating the low end (when it starts writing cache back to disk in the background). 

Something to keep in mind for those who have a large amount of ram installed in their systems; the percentage ratio settings are unlikely to be optimal.

**Comment by gbschenkel** [comment]
> This is a good tip for calculating the low end (when it starts writing cache back to disk in the background).
> 
> Something to keep in mind for those who have a large amount of ram installed in their systems; the percentage ratio settings are unlikely to be optimal.

This is my case, I don't have a lot, but I am using 32GB since 2015, with a Xeon E5-2697v3 (14C/28T). I usually use 8GB on /tmp, to help when compiling and not putting temporary/waste files on the NVMe.

This hangups I can see in two cases outside Steam, which is updating my Distro, when the package has too many small files, and when the files are too large.
